simpleviewinc/keg-hub
# Shared task options and minor lib updates

**Ticket**: [ZEN-463](https://jira.simpleviewtools.com/browse/ZEN-463)

[Semver-Status] minor

## Context

* These changes are part of a bigger ticket to update how docker images are managed. 
* These changes were extracted to to allow for a smaller PR. Because of this; some functionality may not be used in this PR

## Goal

* Allow tasks to share common options
* Improve git lib's fetch functionality
* Fix bug with args-parse array parsing

## Updates

* `repos/args-parse/src/options/checkValueType.js`
  * Fixes a bug when option type is array, but the value is invalid, or does not exist
* `repos/keg-cli/src/constants/docker/volumes.js`
  * Removed unused file
* `repos/keg-cli/src/libs/docker/image.js`
  * Added callback to help when trying to find a docker image
* `repos/keg-cli/src/libs/git/*`
  * Added helper to list remote branches
  * Added helper to print the current git branch commit hash
  * Updated fetch helper to auto-track fetched branches
* `repos/keg-cli/src/tasks/*`
  * Updates some tasks to share common options between them
* `repos/keg-cli/src/utils/task/options/*`
  * Shared task options
  * This allows updating the options in one location, which then updates all tasks that use them

## Testing

**Setup**
* Pull this PR => `keg cli && keg pr 82`
* Install the latest deps => `yarn install`
* Copy the args-parse changes into the keg-cli node_modules
  * `cp ~/keg-hub/repos/args-parse/src/options/checkValueType.js ~/keg-hub/repos/keg-cli/node_modules/@keg-hub/args-parse/src/options/checkValueType.js`
  * This ensures you can test the args-parse changes
* Clear out all your old docker images and cache
  * `keg d clean -f`
  * Be sure to remove the `keg-base` image

**Test 1**
* Run the command `keg base build`
  * This should build the base image without issue
* Run the command `keg components build`
  * This should build the `keg-components` image without issue
* Run the command `keg evf build`
  * This should build the `tap` image without issue

**Test 2**
* Remove all the image images built in **Test 1**
* Run the command `keg components start`
  * This should build the `keg-base` && `keg-components` images without issue
  * It should then start the `keg-components` container
  * Ensure keg-components runs as expected
* Run the command `keg evf start`
  * This should build only the `tap` image without issue, because the base image was built in the last step
  * It should then start the `events force tap` container
  * Ensure the tap events force runs as expected

**Test 3**
* Run the command `keg fetch`
  * This should fetch all branches from the origin remote, and automatically set them to track their remote counterpart

**Test 4**
* Run `keg evf pack run package=docker.pkg.github.com/simpleviewinc/keg-packages/tap:zen-468-pressable-update`
  * This tests to ensure the `docker provider pull` task still works as expected
  * Ensure the package runs as expected

**Test 5**
* Run the command `keg cli test`
  * Ensure all cli tests pass
